Add Attachments Interface

davidsoderberg opened the pull request:
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add so attachment can be sent.
- **What is the current behavior?** (You can also link to an open issue here)
https://github.com/notifirehq/notifire/issues/21
- **What is the new behavior (if this is a feature change)?**
Attachment can be sent.

scopsy commented:
Thinking out loud, what is the best way to incorporate this to all the current providers and what about providers that don't support attachments until we migrate them all. @ComBarnea @davidsoderberg

davidsoderberg commented:
Can we implement a feature property that is default false and then in handler we check if that one is true otherwise we throw an error if there is any attachments in payload?

scopsy commented on packages/core/src/lib/handler/sms.handler.ts:
```diff
@@ -1,11 +1,19 @@
   constructor(private message: IMessage, private provider: ISmsProvider) {}
 
   async send(data: ITriggerPayload) {
+    data.$attachments = data.$attachments?.filter((item) =>
```
```suggestion
    const attachments = data.$attachments?.filter((item) =>
```
I'd prefer if we don't override function params and instead create new const for that here

scopsy commented on packages/core/src/lib/handler/sms.handler.ts:
```diff
@@ -1,11 +1,19 @@
 
   async send(data: ITriggerPayload) {
+    data.$attachments = data.$attachments?.filter((item) =>
+      item.channels?.includes(ChannelTypeEnum.SMS)
```
if no channels passed we should behave as if all channels we're selected  here (same on email)

davidsoderberg replied:
So logic should be something like this:
```
item.channels === undefined ||
        item.channels.length === 0 ||
        item.channels?.includes(ChannelTypeEnum.EMAIL)
```
Did I understand you correctly?

scopsy replied:
`item.channels?.length ? item.channels?.includes(ChannelTypeEnum.EMAIL) : true`

Maybe something like this?

scopsy replied:
And we can add a unit test for this it would be amazing 🚀

davidsoderberg commented:
@scopsy, can we implement a feature property that is default false and then in handler we check if that one is true otherwise we throw an error if there is any attachments in payload?

scopsy commented:
@davidsoderberg I don't think it's neceserry. We can just create another PR. Into this one with the provider changes. And then merge everything together

scopsy commented:
@davidsoderberg I've changed the base of the repo to the `v0.3.0` and added the full task list at #21 for what we need to finish for this. I'll give a hand as well on this. To get this right we need to do some modifications and update the providers. We then can release it on v0.3.0. WDYT?

scopsy commented on packages/core/src/lib/content/content.engine.ts:
```diff
@@ -1,8 +1,12 @@
 import * as Handlebars from 'handlebars';
+import { IAttachmentOptions } from '../..';
```
@davidsoderberg Not sure if this import is ok 😅

davidsoderberg replied:
Woops, auto import messed up 😞

scopsy replied:
Happens to me all the time 😅

davidsoderberg commented:
> @davidsoderberg I've changed the base of the PR to the `v0.3.0` branch and added the full task list at #21 for what we need to finish for this. I'll give a hand as well on it. To get this right we need to do some modifications and update the providers. We then can release it on v0.3.0. WDYT?

I have also realized that this PR creates demands existing providers, sounds great 😄 
I have a solution to warn user of Notifire that the current provider does not support attatchments... 
Do you like to see it or should we look into anther solution for that later? How should I show it to you?

scopsy commented:
You can just add a snippet on the new issue I created for that if that is ok at #104  :) 

I’ll merge this one for now
